aws-lc-rs/src/rsa/oaep.rs

```diff
@@ -0,0 +1,501 @@
+pub struct PrivateDecryptingKey {
+    key: LcPtr<EVP_PKEY>,
+    size: KeySize,
+}
+
+impl PrivateDecryptingKey {
+    fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
+        let size = KeySize::from_evp_pkey(&key)?;
+        Ok(Self { key, size })
+    }
+
+    /// Generate a new RSA private key for use with asymmetrical encryption.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs during the generation of the RSA keypair.
+    pub fn generate(size: KeySize) -> Result<Self, Unspecified> {
+        Self::new(generate_rsa_evp_pkey(size)?)
+    }
+
+    /// Construct a `PrivateDecryptingKey` from the pvoided PKCS#8 (v1) document.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs during deserialization of this key from PKCS#8.
+    pub fn from_pkcs8(pkcs8: &[u8]) -> Result<Self, KeyRejected> {
+        unsafe {
+            let evp_pkey = LcPtr::try_from(pkcs8)?;
+            super::key::validate_rsa_pkey(&evp_pkey)?;
+            Self::new(evp_pkey).map_err(|_| KeyRejected::unexpected_error())
+        }
+    }
+
+    /// Returns the corresponding [`KeySize`].
+    #[must_use]
+    pub fn key_size(&self) -> KeySize {
+        self.size
+    }
+
+    /// Retrieves the `PublicEncryptingKey` corresponding with this `PrivateDecryptingKey`.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs computing the public key.
+    pub fn public_key(&self) -> Result<PublicEncryptingKey, Unspecified> {
+        if 1 != unsafe { EVP_PKEY_up_ref(*self.key) } {
+            return Err(Unspecified);
+        };
+        PublicEncryptingKey::new(LcPtr::new(*self.key)?)
+    }
+
+    /// Decrypts the contents in `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs while decrypting `ciphertext`.
+    pub fn decrypt<'output>(
+        &self,
+        algorithm: &'static EncryptionAlgorithm,
+        ciphertext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified> {
+        let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.key, null_mut()) })?;
+
+        if 1 != unsafe { EVP_PKEY_decrypt_init(*pkey_ctx) } {
+            return Err(Unspecified);
+        }
+
+        configure_oaep_crypto_operation(&pkey_ctx, algorithm.hash(), algorithm.mgf())?;
+
+        let mut out_len = output.len();
+
+        if 1 != indicator_check!(unsafe {
+            EVP_PKEY_decrypt(
+                *pkey_ctx,
+                output.as_mut_ptr(),
+                &mut out_len,
+                ciphertext.as_ptr(),
+                ciphertext.len(),
+            )
+        }) {
+            return Err(Unspecified);
+        };
+
+        Ok(&mut output[..out_len])
+    }
+}
+
+impl AsDer<Pkcs8V1Der> for PrivateDecryptingKey {
+    fn as_der(&self) -> Result<Pkcs8V1Der, Unspecified> {
```

Comment:
Would be good to get some feedback on this `Pkcs8V1Der` type, and this pattern of using `AsDer` trait. Formally on other existing API types we would have used `to_pkcs8` / `to_pkcs8v1` but this was prior to the introduction of the AsDer and corresponding buffer type.

I'm conflicted on trying to deprecate the `pkcs8::Document` type, but it would be nice to distinguish V1 and V2 types in a manner that could make usage with AsDer more compatible. Alternatively we introduce a type alias for Buffer and using generic in that manner, but then that seems rather unnecessary, the AsDer contract is generic and not specific to using a buffer type.

Either way we